### ISSUE TITLE
LabelList - Y value is undefined on content props

### ISSUE THREAD

**Issue by benedekdaniel** [recharts/recharts#3344]
- [x] I have searched the [issues](https://github.com/recharts/recharts/issues) of this repository and believe that this is not a duplicate.

### Reproduction link
[![Edit on CodeSandbox](https://codesandbox.io/static/img/play-codesandbox.svg)](https://codesandbox.io/s/tender-bose-6eh5cs)

### Steps to reproduce
Try simply changing the recharts version from 1.8.5 to 2.3.2 in the package.json.

### What is expected?
It's expected that the y value on the LabelList content props is not undefined but returns a dynamic value relative to the chart's y and x-axis so it always fits on the x axis.

### What is actually happening?
Notice that the custom label component rendered in LabelList doesn't have an y value (it's undefined) in version 2.3.2. After setting the version back to 1.8.5 the noDataLabel chart appears as y gets a value. 

| Environment | Info |
|---|---|
| Recharts | v2.3.2 |
| React | 18.0.0 |
| System | MacOs Monterey version 12.4 |
| Browser | Chrome Version 109.0.5414.119 (Official Build) (arm64) |

**Comment by ckifer** [comment]
Hey there @benedekdaniel 

From what I can test the `y` value breaks as of version 2.0.10 - it is present in `2.0.9`.

Try upgrading to 2.0.9 and in the meantime I'll look into the breaking change.
Upgraded CodeSandbox: https://codesandbox.io/s/kind-lovelace-f0bzs2?file=/src/App.tsx

**Comment by ckifer** [comment]
Diff between versions - https://github.com/recharts/recharts/compare/v2.0.9...v2.0.10

Probably this PR  https://github.com/recharts/recharts/pull/2477

**Comment by ckifer** [comment]
I can't really tell what is happening based on the diff alone - nothing really looks out of place there. 

I do find it odd that you are using `LabelList` to render a bar item.
There might be a better approach for this.

**Comment by ckifer** [comment]
Really confused on why this broke specifically in 2.0.10.

This line: https://github.com/recharts/recharts/blob/master/src/cartesian/Bar.tsx#L195
Has potential to make height `NaN` (makes sense. 300 - undefined = `NaN`). That causes some calculations down the line to use `NaN` and therefore make `y` undefined or `NaN`. Defaulting this to 0... breaks more things because it results in the bars that are `null` being totally full. 

i.e.
![image](https://user-images.githubusercontent.com/25180830/217405843-ef7b4345-746e-4a82-bcf3-bc09fc7b81d7.png)


As for `LabelList` it uses `Label` which does some calculations to determine x and y values here https://github.com/recharts/recharts/blob/7f913f27f35c688e3e3554363c1bf8899d58ce06/src/component/Label.tsx#L187-L266

Those calculations are done using mostly x, y, and height. If `y` or `height` is `NaN` these calculations will also return `NaN`. Can't default to 0 for the same reasons above.

**Comment by ckifer** [comment]
![image](https://user-images.githubusercontent.com/25180830/217412948-3e800d76-59ad-4cba-a2b7-ebc776ced1c3.png)

Got it to work both ways - would recommend taking the custom `shape` approach though, much cleaner imo. 

```jsx
 <Bar
     dataKey="group"
     fill="#8884d8"
     shape={props => {
        if (props.group) return <Rectangle {...props} />;
        return <NoDataBar {...props} />;
     }
 >
```

Then if you want text labels you can do it in the `Bar` shape.

I know where the fix is but still not sure what commit changed it. Will probably be some time before we'll get the fix out.

**Comment by ckifer** [comment]
In draft still but main change is here 

https://github.com/recharts/recharts/pull/3346/files?diff=split&w=0#diff-16e2d1fc138f9202fef0537f5b0559331cf1bf1471acaf3e8005d17a8f0ad54dR194

**Comment by ckifer** [comment]
@benedekdaniel  Should be fixed with the merging of #3346 - Going to close this. Should be fixed on next minor release (tomorrow if all goes well)

**Comment by brightmileDaniel** [comment]
@ckifer Alright thanks, I'll have a look!

**Comment by ckifer** [comment]
Should be fixed in 2.4.0
